text/0000-update_asset_packaging.md

```diff
@@ -0,0 +1,102 @@
+- Feature Name: update_asset_packaging
+- Start Date: 2016-05-10
+- RFC PR:
+- Issue:
+
+# Summary
+[summary]: #summary
+
+Several suggestions have been made to improve the way we handle our assets packaging.
+This RFC attempts to present the various options so that a decision can be made.
+
+# Motivation
+[motivation]: #motivation
+
+Currently, most assets (js/css) are used as gems and managed by bundler.
+A few are included directly in the vendor/ directory due to lack of gems.
+Some of the reasons that we may want to change the way we currently handle this:
+* Some assets are not packaged as gems, so we need to either vendor or package them ourselves.
+* Some asset gems are not updated at the same rate as the upstream source.
+* Each gem dependency must be packaged separately as RPM/DEB.
+* Difficult to integrate newer JS technology such as es2015 or js modules.
+* Assets included in the vendor/ directory need manual updating when new versions are released.
+
+
+# Detailed design
+[design]: #detailed-design
+
+This will depend on the alternative that will be chosen and updated later.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Some reasons to stay with the current solution:
+* No extra work needed.
+* Single package management system - bundler. All dependencies are in one central location, build process is simple.
+* Rails asset pipeline doesn’t require any modifications to handle the current solution.
+
+# Alternatives
+[alternatives]: #alternatives
+
+__Option 1: Stay with current solution__
+
+See above for reasons for and against.
+
+__Option 2: Webpack + npm__
+
+[POC PR](https://github.com/theforeman/foreman/pull/3433)
+
+_Pros:_
+* Live reload development server.
+* Ability to package the code in separate chunks that are only loaded when needed.
```

Comment:
> Many websites don't in fact concatenate everything into one file, but rather a few.

That's true, I'm not sure if there is a consensus on best practices here.  I would imagine that it has to do with the size of your file.

> I don't think we should have 20 different js files for every page - but also going to the other extreme of always serving everything in one file is not the right way in my opinion.

After thinking about this some more I tend to agree.  I think having something like one file for libraries, one for our shared code, and then one per abstract UI components could be okay.  What I think we should avoid is having one per page or per specific UI component (i.e. hosts.js, smart_proxies.js, host-table.js, etc).   Of course I am talking about production mode only, in development such distinctions may make sense.